Alex and Happy Part 1/article.md

```diff
@@ -159,27 +170,53 @@ Lexers are often implemented using Deterministic Finite Automata (FDAs), which y
+The arrow coming from nowhere to the square with 0 is where you start driving, and squares with another square inside are places where you can park your car. In this case, you are not allowed to park it in the square with 0.
 
-State 0 is the initial state where the automaton begins. Suppose that the automaton can stop at states 1, 2, 3, and 4.
+The arrows also have directions in them, guided by your GPS. Your GPS may tell you to drive straight into an `i`, then turn at an `[a-z]`, and then to stop, for example. However, your GPS may also ask you to take illegal turns, for example, to turn at an `@`, which will always lead you to a dead end, where you will be stuck.
+
+To refine the terminology, the city is the _automaton_ (plural: _automata_) represented by a _state transition graph_, each numbered square is called a _state_, and each arrow is called a _transition_.
+When a transition comes from nowhere and into a state, we say that this state is a _start state_, which is state 0 in this example.
+The places were you can park are called _accept states_, i.e., 1, 2, 3, and 4.
+
+The GPS represents the _input_ (string) of the automaton.
+The dead ends represent an implicit _error state_ present in every DFA, which you can think of like a hidden state number 5, with all invalid transitions from every state to it.
+
+To simplify matters, on ambiguous transitions such as `i` and `[a-z]`, assume that the one with a single letter takes precedence.
 
-If we want to check the expression "int", we will follow this path: 0 → 1 → 2 → 4.
+For example, to check the expression "int", we will follow this path: 0 → 1 → 2 → 4.
 
 Now, we need to attribute meanings to each state so that this automaton can be useful, and we will do so like this:
 * 0: Error.
 * 1: Identifier.
 * 2: In.
 * 3: If.
 * 4: Identifier.
+* 5 (implicit): Error.
 
 Since we stopped at 4, the result is that "int" is an identifier, so we get a token that is `Identifier "int"`.
-The resulting token would be 'In' if we stopped at 2 instead (for the expression "in").
+The resulting token would be `In` if we stopped at 2 instead (for the expression "in").
 
-
-What if we consumed no input and stopped at state 0, or the input was something like "Foo123" where there aren't valid transitions?
+What if we consumed no input and stopped at state 0, or the input was something like "Foo123" where we'd reach the implicit state 5?
 In such cases, we halt and indicate a lexical error.
 
-We won't write our lexical analyzer from scratch. Instead, we will use Alex to create a list of tokens, which Happy will parse.
+As started previously, we won't write our lexical analyzer from scratch. Instead, we will use Alex to generate a lexer that creates a stream of tokens, which the Happy-generated parser will parse.
+
+### Regular expressions
+
+Alex uses regular expressions (regexes) to match patterns, we'll keep it pretty simple and explain some of them which we'll use.
```

Comment:
Ah, oops, I remember Thompson's algorithm from my university classes (although I didn't know about the other two people in the name). I had a brain fart to write match instead of define. Thanks for correcting me.